### ISSUE TITLE
Generate decoder/encoder for a given msg-type

### ISSUE THREAD

**Issue by opvasger** [opvasger/elm-dev-tools#19]
One of the biggest issues with this module is that the it requires the consumer to maintain encoders/decoders for their applications msg-type.

I've experimented with different solutions to this problem:
- Monkey-patching compiled elm modules (god forbid!)
- Code-generation based on static-analysis
- Asking Evan to provide an API

I think the right approach at this point is code-generation. One of the great benefits of generating the encoders/decoders is that the error-messages are great and easy to work with. Coercing values from/to Js/Elm is worked in a lot of ways, but even after monkey-patching the functionality into the program, errors remained painful to deal with.

I imagine building a CLI-tool that, given a type-definition and a module-path can produce a module with an encoder and a decoder for the given type. If types are opaque or unreachable in some other way, it should give an accurate error-message.

**Comment by MartinSStewart** [comment]
With the code generation solution, wont this be unable to decode some types? For example in elm-exploration/web-gl the Texture type can only be created by the elm runtime.

**Comment by opvasger** [comment]
> For example in elm-exploration/webgl the Texture type can only be created by the elm runtime.

Yes, unfortunately some libraries rely on messages carrying opaque values.

Is it fair to rephrase your question to:

    Does an idiomatic Elm `Msg`-definition ever need its
    constructors to carry opaque value(s)?

I've added "idiomatic" as I wouldn't mind the debugger helping me improve my code!

1. Modules installed with the package-manager often have exposed type-constructors that will be harder to analyse, but I'm confident it can be done.
2. `elm/file` has an opaque `File`-type. Just like with `elm/webgl` and `Texture`, I don't know how to generate code for it's en/decoder.

I'm very reluctant to revisit magic APIs or monkey-patched compiler-output. Even if `elm-devtools` can't support every application right away, the ecosystem could adjust to provide better support?

Any thoughts? :sunny:

**Comment by MartinSStewart** [comment]
```
Does an idiomatic Elm `Msg`-definition ever need its
constructors to carry opaque value(s)?
```
That's fair.  I think the answer to that question is no.  There are cases where you'd have to choose between idiomatic messages and messages that have no opaque types.


In the context of automatically generating decoders and encoders that don't rely on any compiler magic, there seem to be three categories of message constructors.

1. Constructors that can be automatically decoded and encoded.  All their parameters (if any) are type aliases, primitives, or custom types whose constructors all fit into category 1.

2. Constructors that need to be manually decoded and encoded.  Any constructors with parameters that are opaque but have functions made for constructing and deconstructing them end up here.

3. Constructors that cannot be decoded and encoded. Constructors with parameters that aren't exposed and don't have functions that can construct and deconstruct them are here.


Looking at a project I'm currently working on, in my message type there are 4 constructors with parameters that are in category 3. `Result WebGL.Texture.Error WebGL.Texture.Texture`, `Keyboard.Msg` (from ohanhi/keyboard), `Browser.Dom.Element`, and `File.File`.  I don't think there's any way to modify these so that they can be decoded and encoded (regardless of how idiomatic the code is or isn't).

Of the remaining constructors, 30 are in category 1 and 19 are in category 2.  The category 1 stuff is mostly constructors that have no parameters or a single primitive parameter. For category 2, I have several constructors that have `Point2d` as a parameter (from ianmackenzie/elm-geometry). I could swap these to instead be `{ x : Float, y : Float }` and then they'd be in category 1, but I think `Point2d` is idiomatic because using `Point2d` gives more meaning, i.e. it's not a vector, or a size, it's a position.

This is for a game so maybe my use case is unusual and people writing webpages will have better luck.  I think it's worth looking into.  Eitherway, in my case, I don't think automatic code generation for message constructors can work without compiler magic.

All that said, whether or not message constructors can get decoded/encoded only affects exporting and importing history correct? Time travel debugging and seeing the contents of a message can still work without it right?

**Comment by opvasger** [comment]
>   That's fair. I think the answer to that question is no. There are cases where you'd have to choose between idiomatic messages and messages that have no opaque types.

When I say "idiomatic", I just mean to rule out crazy cases that aren't practical or necessary. An example of an idiomatic `Msg`-constructor is `DecodeFile File` based on `elm/file`. In that sense, I'd say the answer is yes - you do need to work with messages that carry opaque-types that cannot be converted. Maybe `elm/file` could get a minor version-bump to expose functionality to do this? I don't see why not at this point, but I should investigate to get more context.

I guess this turns the question into:

    Does an idiomatic Msg-type in Elm ever need to carry opaque values that could never be 
    en/decoded?

Based on your feedback, I think `Browser.Dom.Element` is probably the most problematic? I'll also investigate if this can somehow be dealt with :sunny:

> All that said, whether or not message constructors can get decoded/encoded only affects exporting and importing history correct? Time travel debugging and seeing the contents of a message can still work without it right?

Yes, turning messages from/to JSON is required for:
1. Downloading/Uploading the current session
2. Caching the current session to local-storage, so you get 100% reliable hot-reloading between browser-reloads
3. Printing individual messages within the UI for inspection

**Comment by MartinSStewart** [comment]
`Based on your feedback, I think Browser.Dom.Element is probably the most problematic?` For my use case, `Texture` is the most problematic, both in terms of importance and how difficult I imagine it would be to convert it to json.  I imagine `Browser.Dom.Element` is the most problematic for general use cases though.

**Comment by opvasger** [comment]
You're absolutely right 😄oops.

`File` still makes the case relevant tho. Let's see what can be done!

**Comment by MartinSStewart** [comment]
Is the idea to ask the maintainer of elm/file (which I'm guessing is Evan) if they can add a encoder and decoder to the API?

**Comment by opvasger** [comment]
Yes, I'd like to include it when I have a few other things ready for him regarding devtools. He has already given some advice, and I'll eventually include it in a Discourse-thread when `DevTools.Browser` is ready.

not necessarily en/decoders, but whatever options makes sense for the this module and the ecosystem. I want full support for all Elm apps, and there's a lot to do! :sunny:
